### ISSUE TITLE
(PACER bug) Text-only entries have no unique URL

### ISSUE THREAD

**Issue by calvinli** [calvinli/pacerrssscraper#14]
This is actually a bug in PACER, but it affects us, so I'm posting it here.

The RSS feeds include "guid" elements, e.g. https://ecf.wawd.uscourts.gov/cgi-bin/DktRpt.pl?241761&129 . For text-only entries, this is the only link available, and so is the one we use in notifications (e.g., tweets).

These links don't actually work. If you follow that link, it leads to an empty Docket sheet page, with the "Case number" field blank (or potentially even wrong*). Obviously that makes the link entirely useless (*so why is it in the RSS feed???*).  **Interestingly**, if you strip the sequence number off (the ampersand and all that follows), it does work (it leads to the docket sheet request form for that case), but of course then you lose all information about what docket entry you were looking for.

(*There is another wrinkle here, involving the fact that if you do a case search PACER keeps track of what case you're on via a `CASE_NUM` *cookie* ?!? which is then used to auto-fill the case number field on subsequent requests, and it seems to never get cleared out, but this is getting far too complicated already.)

This has some ancillary effects: since we strip off the sequence number (so the links actually lead somewhere useful), it is possible that in the event of multiple new text-only docket entries in the same case with the same title, we will end up submitting exact duplicate tweets, which is a Twitter API error (you may not send duplicate tweets, it leads to a HTTP 403). In such a case, though, we actually *do* want to submit one tweet for each entry.  (This happened today in *State of Washington v. Trump*, WAWD 17-cv-00141, with three consecutive text-only entries named "Order on Application for Leave to Appear Pro Hac Vice".)

There is a deeper root cause here: text-only entries *have* docket numbers, but they cannot be seen from `qryDocument.pl` the same way that PDF entries can. The only way to see text-only entries is to purchase the entire docket (at least, for the time period in which you think the text-only entry was filed). The other is that PACER internally tracks documents via an internal case_id (sequential within the district, iirc) and a sequence number de_seq_num (which I think increments for each page of each document on the docket), but externally of course documents are referenced by the district/year/type/serial case number (e.g. WAWD 17-cv-00141) and a document number.

Example: here is the link given in the RSS feed for WAWD 17-cv-00141 \#25: https://ecf.wawd.uscourts.gov/doc1/19717305398?caseid=241761&de_seq_num=111. Neither the external case number nor the document number appear in this URL. Also, the query parameters are ignored entirely, the URL works just fine with them removed. The field `19717305398` appears to be four digits of unclear meaning, then a field that PACER calls `dlsid`, which is sequential among all documents in the district but neither increases by one each time nor by the number of pages. It is this `dlsid` that PACER uses to retrieve the document, and you need to know that `dlsid` in order to get a link to that document, and it cannot be derived from the case number and document number. That being said, PACER is clearly able to derive both the case number and document number from this `dlsid`, because it appears on the PACER receipt shown before you can view the document.

Basically the point is that PACER has some *serious* architectural issues in how it is organized.

**Comment by PlainSite** [comment]
This may be totally unrelated, but another bug that really annoys me is the situation where you have a multi-party criminal case with several similar docket numbers that are incremented for each defendant. (1:08-cr-03923-ABC, 1:08-cr-03923-ABC-1, 1:08-cr-03923-ABC-2, etc.) If you click on a document link *provided by PACER* for -1, you may get an error saying something like "cannot find case_num xxxxxx", where xxxxxx is the case number of the primary docket the document appears to be actually linked to. Replacing the case_num in the URL with the one that it somehow can't find then will work. But if it knows that already, why can't it find the document?

Because, as you say above in a commendable and understated manner, "PACER has some serious architectural issues in how it is organized."
